### ISSUE TITLE
CSRFError on source interface with HTTPS

### ISSUE THREAD

**Issue by redshiftzero** [freedomofpress/securedrop#3333]
## Description

The source interface is not functional for _some_ instances using HTTPS (though not all, as there are some sites that are using HTTPS without issue). 

I'm writing out what I know so far based on talking to impacted sites directly and https://forum.securedrop.club/t/incorrect-docs-for-https-on-the-source-interface-issues-after-configuration/657/3 

## Steps to Reproduce

I do not have full steps to reproduce. This needs further investigation. There is a ticket for end-to-end testing of the HTTPS flow at #3297.

So far:

1. Apply HTTPS on source interface 
2. Try to submit a document

## Expected Behavior

A source is able to submit a document and all is well.

## Actual Behavior

You are logged out: 

![image](https://user-images.githubusercontent.com/7832803/39488472-49ccc286-4d37-11e8-9fee-6df9940efc18.png)

## Comments

I can report from my debugging with a site that experienced this issue that the issue producing the logout was a `CSRFError`. The `errorhandler` in `securedrop/source_app/__init__.py` produced the observed behavior:

```
    @app.errorhandler(CSRFError)
    def handle_csrf_error(e):
        msg = render_template('session_timeout.html')
        session.clear()
        flash(Markup(msg), "important")
        return redirect(url_for('main.index'))
```

The specific exception causing the CSRF validation failure was `400 Bad Request: The referrer header is missing`. This Referrer checking is due to [`WTF_CSRF_SSL_STRICT` which is by default on](https://flask-wtf.readthedocs.io/en/stable/config.html). 

Indeed, in `/etc/apache2/sites-available/source.conf`, replacing:

```
Header set Referrer-Policy "no-referrer"
```

with 

```
Header set Referrer-Policy "same-origin"
```

resolves the issue as reported by a public SecureDrop user [here](https://forum.securedrop.club/t/incorrect-docs-for-https-on-the-source-interface-issues-after-configuration/657/4).

**Comment by conorsch** [comment]
As part of #3299, took a gander at the HTTPS config with self-signed certs in prod VMs. Can confirm:

1. The reports of a broken Source Interface over HTTPS. It loads fine, but actually submitting documents doesn't work (due to the CSRF settings @redshiftzero explains above). 
2. The `Referrer-Policy` tweak indeed resolves the problem. After a `sudo service apache2 reload`, I was able to interact with the Source Interface normally again, including making submissions.

Great explanation above, @redshiftzero. Let's get the fix into 0.7!

**Comment by redshiftzero** [comment]
There are two options that would _technically_ resolve this issue: setting `WTF_CSRF_SSL_STRICT` to False, and setting the `Referrer-Policy` to `same-origin`, so I'm documenting the rationale of recommending this to prod instances here. 

### Option 1: Set `WTF_CSRF_SSL_STRICT` to False

The`WTF_CSRF_SSL_STRICT` option is currently set by default to `True` (Strict Referer Validation) for HTTPS requests, and setting it to False *would* resolve the situation. However, the option is in place for security reasons i.e. to ensure that the `Referrer` is from the correct origin prior to validating the CSRF token. As such I would prefer _not_ to disable this strict validation unless absolutely necessary (for an older but good reference as background, see [this paper](http://www.adambarth.com/papers/2008/barth-jackson-mitchell-b.pdf) on validating the Referer header as a defense against CSRF attacks). 

### Option 2: Set `Referrer-Policy` to `same-origin`

The other option is to change the `Referrer-Policy`. Currently we have the `Referrer-Policy` on the source interface set to `no-referrer`. The intention here was not to leak that a user is coming from a SecureDrop instance when they navigate to a site that we link to on the interface. Overall a good thing. While this makes sense, Tor Browser [will not send Referers from onion service websites](https://trac.torproject.org/projects/tor/ticket/9623) anyway and as such the change added in #2753 was unnecessary (indeed, this was [noted during review](https://github.com/freedomofpress/securedrop/pull/2753#pullrequestreview-83995047)). For defense in depth (e.g. in case there was a regression in the behavior in Tor Browser), instead of removing the `Referrer-Policy` entirely, we can set it to `same-origin`: we still have a restricted `Referrer-Policy` that won't provide referers to third parties _and_ CSRF validation will work. 

So since option 2 has no negative implications, it is the best option. 

### Why did this production bug happen

We merged #2753 _after_ we added support for HTTPS on the source interface in #1605, so this bug was _not_ due to lack of QA prior to adding HTTPS support, but was a regression due to lack of automated testing of the HTTPS cert flow on the source interface. As such, the way to avoid this in the future is to improve automated testing of HTTPS on the source interface (see #2036 which I just added to the 0.8 milestone).
